qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+
+    def visit_AlignmentKind(self, node: AlignmentKind):
+        """Get parameters from block's ``AlignmentKind`` specification."""
+        for param in node._context_params:
+            self.visit(param)
+
+    # Mid layer: Get parameters from instructions
+
+    def visit_Call(self, node: instructions.Call):
+        """Get parameters from ``Call`` instruction.
+
+        .. note:: ``Call`` instruction has a special parameter handling logic.
+            This instruction separately keeps parameters and program.
+        """
+        for parameter in node.parameters:
+            self.parameters.add(parameter)
+
+    def visit_Instruction(self, node: instructions.Instruction):
+        """Get parameters from general pulse instruction.
+
+        .. note:: All parametrized object should be stored in the operands.
+            Otherwise parameter cannot be detected.
+        """
+        for op in node.operands:
+            self.visit(op)
+
+    # Lower layer: Get parameters from operands
+
+    def visit_Channel(self, node: channels.Channel):
+        """Get parameters from ``Channel`` object."""
+        if isinstance(node.index, ParameterExpression):
+            self._parse_parameter_expression(node.index)
+
+    def visit_ParametricPulse(self, node: ParametricPulse):
+        """Get parameters from ``ParametricPulse`` object."""
+        for op_value in node.parameters.values():
+            if isinstance(op_value, ParameterExpression):
+                self._parse_parameter_expression(op_value)
+
+    def visit_Waveform(self, node: Waveform):
+        """Get parameters from ``Waveform`` object.
+
+        .. node:: No parameter can be assigned to ``Waveform`` object.
+        """
+        pass
+
+    def generic_visit(self, node: Any):
+        """Get parameters from object that doesn't belong to Qiskit Pulse module."""
+        if isinstance(node, ParameterExpression):
+            self._parse_parameter_expression(node)
+
+    def _parse_parameter_expression(self, param_expr: ParameterExpression):
+        """A helper function to get parameters from parameter expression."""
+        for parameter in param_expr.parameters:
+            self.parameters.add(parameter)
+
+
+class ParameterManager:
+    """Helper class to manage parameter objects associated with arbitrary pulse programs.
+
+    This object is implicitly initialized with the parameter object storage
+    that stores parameter objects added to the parent pulse program.
+
+    Parameter assignment logic is implemented based on the visitor pattern.
+    Instruction data and its location are not directly associated with this object.
+    """
+    def __init__(self):
+        """Create new parameter table for pulse programs."""
+        self._parameters = set()
+
+    @property
+    def parameters(self) -> Set:
+        """Parameters which determine the schedule behavior."""
+        return self._parameters
+
+    def is_parameterized(self) -> bool:
```

Comment:
Like I commented elsewhere, this seems useful! Seems like I should be able to say `instruction.is_parameterized` by using this underlying method